### PR TITLE
Changing the string null check to improve performance.

### PR DESCRIPTION
**Pull request by sann3**
isEmpty() gives **20-40%** more performance than "".equals().

String equals source code, 
```
public boolean equals(Object anObject) {
        if (this == anObject) {
            return true;
        }
        if (anObject instanceof String) {
            String aString = (String)anObject;
            if (coder() == aString.coder()) {
                return isLatin1() ? StringLatin1.equals(value, aString.value)
                                  : StringUTF16.equals(value, aString.value);
            }
        }
        return false;
    }
```

Operator instanceOf, class casting, 3 if statements, and all this is just to start the check if the string is actually empty. 

Now compare it with String.isEmpty():
```
public boolean isEmpty() {
    return value.length == 0;
}
```

If we increase the performance in the framework, automatically multiple projects get the benefit.

#hacktoberfest

### CODE REVIEW

**Review by snicoll — CHANGES_REQUESTED** [review]
Thanks for the PR. The concrete impact of the change can be argued as most of those aren't hot code paths. I've added a few comments, can you please review?

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/amqp/RabbitProperties.java [hunk 1]
```diff
@@ -285,7 +285,7 @@ public String determineVirtualHost() {
 	}
 
 	public void setVirtualHost(String virtualHost) {
-		this.virtualHost = "".equals(virtualHost) ? "/" : virtualHost;
+		this.virtualHost = (virtualHost == null || virtualHost.isEmpty()) ? "/" : virtualHost;
```
[the remark this answers is not in the document]
Comment (sann3) [reply spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/amqp/RabbitProperties.java:288]: Yes, I will update it hasLength method.

File: spring-boot-project/spring-boot-cli/src/main/java/org/springframework/boot/cli/command/shell/CommandCompleter.java [hunk 2]
```diff
@@ -74,7 +74,7 @@ public int complete(String buffer, int cursor, List<CharSequence> candidates) {
 		int completionIndex = super.complete(buffer, cursor, candidates);
 		int spaceIndex = buffer.indexOf(' ');
 		String commandName = ((spaceIndex != -1) ? buffer.substring(0, spaceIndex) : "");
-		if (!"".equals(commandName.trim())) {
+		if (commandName != null && !commandName.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (sann3) [reply spring-boot-project/spring-boot-cli/src/main/java/org/springframework/boot/cli/command/shell/CommandCompleter.java:77]: Yes, I will update it with hasText method.

File: spring-boot-project/spring-boot-tools/spring-boot-maven-plugin/src/intTest/projects/run-jvm-system-props/src/main/java/org/test/SampleApplication.java [hunk 3]
```diff
@@ -32,7 +32,7 @@ public static void main(String[] args) {
 			throw new IllegalStateException("property1 system property mismatch (got [" + property1 + "]");
 		}
 		String property2 = System.getProperty("property2");
-		if (!"".equals(property2)) {
+		if (property2 != null && !property2.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (sann3) [reply spring-boot-project/spring-boot-tools/spring-boot-maven-plugin/src/intTest/projects/run-jvm-system-props/src/main/java/org/test/SampleApplication.java:35]: Yes, here it is not about performance, its about best practise.

**Comment by snicoll** [comment]
@sann3 thanks for the follow-up and for making your first contribution to Spring Boot.

**Pull request closed without merging**